libraries/chain/include/eosio/chain/kv_context.hpp

```diff
@@ -0,0 +1,65 @@
+#pragma once
+
+#include <eosio/chain/name.hpp>
+#include <eosio/chain/kv_config.hpp>
+#include <memory>
+#include <stdint.h>
+
+namespace chainbase {
+class database;
+}
+
+namespace eosio { namespace chain {
+
+   class apply_context;
+
+   inline constexpr name kvram_id  = N(eosio.kvram);
+   inline constexpr name kvdisk_id = N(eosio.kvdisk);
+
+   enum class kv_it_stat {
+      iterator_ok     = 0,  // Iterator is positioned at a key-value pair
+      iterator_erased = -1, // The key-value pair that the iterator used to be positioned at was erased
+      iterator_end    = -2, // Iterator is out-of-bounds
+   };
+
+   struct kv_iterator {
+      virtual ~kv_iterator() {}
+
+      virtual bool       is_kv_chainbase_context_iterator() const                                       = 0;
+      virtual kv_it_stat kv_it_status()                                                                 = 0;
+      virtual int32_t    kv_it_compare(const kv_iterator& rhs)                                          = 0;
+      virtual int32_t    kv_it_key_compare(const char* key, uint32_t size)                              = 0;
+      virtual kv_it_stat kv_it_move_to_end()                                                            = 0;
+      virtual kv_it_stat kv_it_next()                                                                   = 0;
+      virtual kv_it_stat kv_it_prev()                                                                   = 0;
+      virtual kv_it_stat kv_it_lower_bound(const char* key, uint32_t size)                              = 0;
+      virtual kv_it_stat kv_it_key(uint32_t offset, char* dest, uint32_t size, uint32_t& actual_size)   = 0;
+      virtual kv_it_stat kv_it_value(uint32_t offset, char* dest, uint32_t size, uint32_t& actual_size) = 0;
+   };
+
+   struct kv_resource_manager {
+      void     update_table_usage(int64_t delta) { return _update_table_usage(*_context, delta); }
+      apply_context* _context;
+      uint64_t       billable_size;
```

Comment:
It can be const, but should it be?  It doesn't make a big difference either way, but I usually don't make non-static data members const, as it prevents assignment.

Reply:
No, never mind. Just keep as is.